src/Microsoft.ML.DataView/TypeManager.cs

```diff
@@ -0,0 +1,59 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using Microsoft.ML.Internal.DataView;
+
+namespace Microsoft.ML.Data
+{
+    public static class TypeManager
+    {
+        // Types have been used in ML.NET type systems. They can have multiple-to-one type mapping.
+        // For example, UInt32 and Key can be mapped to uint. We enforce one-to-one mapping for all
+        // user-registered types.
+        private static HashSet<Type> _notAllowedRawTypes;
+        private static ConcurrentDictionary<Type, DataViewType> _rawTypeToDataViewTypeMap;
+        private static ConcurrentDictionary<DataViewType, Type> _dataViewTypeToRawTypeMap;
+
+        /// <summary>
+        /// Constructor to initialize type mappings.
+        /// </summary>
+        static TypeManager()
+        {
+            _notAllowedRawTypes = new HashSet<Type>() {
+                typeof(Boolean), typeof(SByte), typeof(Byte),
+                typeof(Int16), typeof(UInt16), typeof(Int32), typeof(UInt32),
+                typeof(Int64), typeof(UInt64), typeof(string), typeof(ReadOnlySpan<char>)
+            };
+            _rawTypeToDataViewTypeMap = new ConcurrentDictionary<Type, DataViewType>();
+            _dataViewTypeToRawTypeMap = new ConcurrentDictionary<DataViewType, Type>();
+        }
+
+        public static DataViewType GetDataViewTypeOrNull(Type type)
+        {
+            if (_rawTypeToDataViewTypeMap.ContainsKey(type))
+                return _rawTypeToDataViewTypeMap[type];
+            else
+                return null;
+        }
+
+        public static Type GetRawTypeOrNull(DataViewType type)
+        {
+            if (_dataViewTypeToRawTypeMap.ContainsKey(type))
+                return _dataViewTypeToRawTypeMap[type];
+            else
+                return null;
+        }
+
+        public static void Register(Type rawType, DataViewType dataViewType)
+        {
+            if (_notAllowedRawTypes.Contains(rawType))
+                throw Contracts.ExceptParam(nameof(rawType), $"Type {rawType} has been registered as ML.NET's default type. " +
+                    $"so it can't not be registered again.");
+            if (_rawTypeToDataViewTypeMap.ContainsKey(rawType))
```

Comment:
>_rawTypeToDataViewTypeMap [](start = 16, length = 25)

this is not atomic.  Prefer regular dictionaries and locks.

Reply:
+1

---
In reply to: [284939889](https://github.com/dotnet/machinelearning/pull/3263#discussion_r284939889) [](ancestors = 284939889)